File: LeetCode/src/dp/DecodeWays.java
package dp;

public class DecodeWays {
    public int numDecodings(String s) {

        int n = s.length();
        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = s.charAt(0) == '0' ? 0 : 1;

        for (int i = 2; i <= n; i++) {
            int num = s.charAt(i - 1) - '0';
            if (num > 0 && num <= 9) {
                dp[i] = dp[i - 1];
            }

            int pre = s.charAt(i - 2) - '0';
            int total = pre * 10 + num;
            if (total >= 10 && total <= 26) {
                dp[i] += dp[i - 2];
            }
        }

        return dp[n];
    }

    public static void main(String[] args) {
        DecodeWays solution = new DecodeWays();

        String s1 = "12";
        System.out.println(solution.numDecodings(s1)); // Output: 2

        String s2 = "226";
        System.out.println(solution.numDecodings(s2)); // Output: 3

        String s3 = "0";
        System.out.println(solution.numDecodings(s3)); // Output: 0

        String s4 = "06";
        System.out.println(solution.numDecodings(s4)); // Output: 0

        String s5 = "2101";
        System.out.println(solution.numDecodings(s5)); // Output: 1
    }
}

File: LeetCode_Python/dp/DecodeWays.py
class DecodeWays:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        dp = [0] * (n + 1)
        dp[0] = 1
        dp[1] = 1 if s[0] != "0" else 0

        for i in range(2, n + 1):
            one = int(s[i - 1])
            if 0 < one <= 9:
                dp[i] += dp[i - 1]

            two = int(s[i - 2 : i])
            if 10 <= two <= 26:
                dp[i] += dp[i - 2]

        return dp[n]


# Time Complexity: O(n)
# Space Complexity: O(n)

if __name__ == "__main__":
    solution = DecodeWays()

    print("Expected Output: 2")
    print("Actual Output:", solution.numDecodings("12"))

    print("Expected Output: 3")
    print("Actual Output:", solution.numDecodings("226"))

    print("Expected Output: 0")
    print("Actual Output:", solution.numDecodings("0"))
